contributors/design-proposals/live-and-inplace-vertical-scaling.md

```diff
@@ -0,0 +1,325 @@
+Resizing running containers of a pod in place could avoid that additional load.
+
+Conversely, resizing without restart may not be the right option for all applications and the right option may also be resource dependent.
+For example, an application might (or have the means to) assess resources available to it only at its initialization time and be unable to adjust its usage if those are dynamically changed.
+Or it might be able to adjust to changes in certain resources (most can for CPU) but not for others (some application have particular difficulty freeing memory they have taken up).
+So we need the means to express and implement the best approach for resizing for each workload and resource.
+
+# Use Cases
+
+* After deploying a service/application through `StatefulSet`, I want to resize, by giving a updated pod spec with new resource sizes, the resources (for now, CPU and memory) allocated to each pod of the statefulset without restarting the pods.
+
+* A (vertical) pod auto-scaler (c.f. [a related issue](https://github.com/kubernetes/features/issues/21)) can utilize this feature, while currently it does resize a pod with restart only.
+
+# Objectives
+
+1. Enable live and in-place resource resizing on a pod.
+2. Add support for live and in-place resource resizing in `StatefulSet` controller.
+
+# API and Usage
+
+To express the policy for resizing in a pod spec, we introduce resource attribute `resizePolicy` with the following choices for value:
+* RestartOnly. (the current behavior, default)
+* LiveResizeable.
+
+This attribute will be available per resource (such as cpu, memory) and so is adequate to indicate whether the workload can handle and prefer a change in each resource’s allocation for it without restarting.
+With potentially multiple containers and multiple resizeable resources for each in a Pod, the response to an update of the pod spec will be determined by the a precedence order among the attribute values with RestartOnly dominating LiveResizeable, i.e., if two resources have been resized in the update to the spec and one of them has a policy of RestartOnly then the pod would be restarted to realize both updates.
+
+We can optionally introduce an action annotation `resizeAction` with following choices for value:
+* Restart. (default)
+* LiveResize.
+* LiveResizePreferred.
+
+If used, this would be included as part of the patch or appropriate update command providing the spec update for the resize.
+It would indicate the preference of user at the time of resize.
+Specifically, Restart for `resizeAction` would indicate the pod be restarted for the corresponding resizing of resource(s), LiveResize would indicate the pod not be restarted the resize be realized live, and LiveResizePreferred would indicate that the resize be realized preferrably live but if that fails for any reason to accomplish it with a restart.
```

Comment:
LiveResize as written is not realizable.  It's always possible that a resize fails for unforeseen reasons, and there has to be a fallback.  The fallback can be a hard restart or a full pod kill, but we can't guarantee a resize-in-place, and our API doesn't really allow kubelet to NAK an update operation (except by killing).